### DISCUSSION TITLE
Is it possible to add TextStamp without signature ?

### DISCUSSION THREAD

**Discussion by pulse-mind** [MatthiasValvekens/pyHanko#194] [Q&A]
Hello,

First of all, thank you for the job !!
I'd like to know if it is possible to add a TextStamp on the PDF without signature. I would use the Incremental Save to not break the existing signatures validity. 
And I'd like to add more than one stamp. 

Is that possible ? If yes do you have some example in the documentation ? 

Thanks a lot.

**Comment by MatthiasValvekens** [comment]
Yes and no. PyHanko includes the option to apply stamps to files; [this is the most high-level entrypoint in the API to do so](https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko.stamp.html#pyhanko.stamp.text_stamp_file). Multiple stamps require more code, but if you look at the implementation of that function, it's actually quite trivial to adapt: it's a simple matter of calling `stamp.apply(...)` multiple times.

In fact, if you go all the way back to the first couple of commits, you'll see that it's been in there from the very beginning :). The focus shifted towards signature-related features very quickly afterwards, so the stamp function has remained kind of primitive, but the essentials are mostly there.

>  I'd like to know if it is possible to add a TextStamp on the PDF without signature. I would use the Incremental Save to not break the existing signatures validity.

This is a different story. The good news is that the stamping API also uses incremental saving by default. The less good news is that modifying page content after signing (even incrementally) will cause any serious PDF signature validator to invalidate all prior signatures anyway. PyHanko's built-in validator would do so too, for what it's worth.

There _might_ be a workaround if you tweak the stamping function to generate stamp annotations instead of direct page content. Some validators would accept that---pyHanko wouldn't---but it's still not a good idea.

So, long story short, I suspect we have an [XY problem](https://en.wikipedia.org/wiki/XY_problem) on our hands :). What's your use case? Maybe there's a better way to achieve what you want.

**Comment by pulse-mind** [comment]
Hi Matthias

Thank you for your answer. 
You're right always going back to the use case :). 

Use case : 
Many user will sign a document, so one after one. 
For each user we have the signature (including the visible signature), and we have initials, legal text,...
I was thinking to manage all these additional "text" as annotation using incremental save. Because annotations seems to be accepted when validating a document with signature. I've seen that it is what DocuSign is doing by adding Stamp Annotation. So the process would be : 
* Upload the PDF
* Add annotations for User 1 (using incrSave)
* Add signature + visible signature (using incrSave)
* Add annotations for User 2 (using incrSave)
* ...


I've done many prototypes with others free librairies but some are not managing incremental save, some are not managing annotation. The only one I found that match the requirements and that was working well is pyMuPDF but in some case the annotation is breaking the Acrobat validation which is what most of lambda users are using for checking Signatories.
I tried pdfBox (Java) and it works too, but our code base is in Python. I can find a solution... but...

**Reply by MatthiasValvekens** [reply]
Hi @pulse-mind,

Sorry for sitting on this for so long, I've been busy.

Right, this is a somewhat common problem, and there's no super-clean solution. From a purely technical PoV, this whole notion of having to initial every page is a holdover from paper-based workflows that doesn't really make sense in the digital signing world. As soon as you start modifying the visual presentation of the page content in (essentially) arbitrary ways, you end up with signatures that are _very_ difficult to validate consistently, or you have to compromise on security ([this blog post](https://itextpdf.com/blog/itext-news-technical-notes/attacks-pdf-certification-and-what-you-can-do-about-them) is a good starting point on where the pain points are).

Anyway, we techies don't make the rules, and in some verticals people are still stuck in the past ;). Representing the kind of workflow you describe in PDF is actually a matter of some controversy even among experts. There's no real consensus on what the "right" way to do it is, but here are some general principles:

- Move as much data as possible into the signature metadata instead of page content (e.g. legal disclaimers can go into the **Reason** field, for example, or even the signature policy descriptor using `sp-user-notice` if you're willing to get your hands a little dirty).
- If you absolutely have to put something on the page in an already-signed document, put it in an annotation. If done right, this should pass muster in most end-user viewers (Acrobat/Foxit/...). FWIW, pyHanko's validator will reject such changes (by design).
- Use incremental saving (duh).

So, the million dollar question: can pyHanko do these things? Yes. Are they effortless to implement? Other than the incremental saving part, not at the moment. Reasons include the following:

 - pyHanko only offers a "nice" API to generate rudimentary signature appearances. It's not difficult to repurpose that code to generate other annotations, but it requires using the low-level PDF manipulation API + some knowledge of the PDF spec.
 - If you want to go the policy route: proper signature policy integration is on the roadmap, but it's blocked by ongoing work on the validation end. That doesn't mean you can't use signature policies, but it does mean that you'll have to [write the plumbing code yourself](https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko.sign.signers.pdf_cms.html#pyhanko.sign.signers.pdf_cms.Signer.signed_attrs). Again, this isn't super hard, but it requires effort.

Anyway, sorry that I don't have a good answer here, but this kind of workflow is intrinsically hard to represent in PDF in an interoperable manner.
